# HackBIo_internship
# stage 1 assay 

## Steps in creating calculate_protien_weight_kDa function
The first step in calculating the total weight of protein is the naming format. 
To solve this issue, I created an `if/else` statement to handle One_letter_code and Three_letter_code naming formats separately. 
The second step is using a `for loop` to automate adding up the weight of each amino acid in the protein name
to a variable called `protein_wt`.
last step is unit transformation to KDa.


# stage 2 tutorial
## Part One
### Heatmap of top differentially expressed genes between HBR and UHR 
![Figure 1a](plots/figure1a.png)

### Vln plot of gene expression 
![Figure 1b](plots/figure2a.png)

## Part Two
### Scatter plot of breast cancer tissue radius vs. texture
![Figure 1c](plots/figure2c.png)

### Correlation heatmap
![Figure 1d](plots/figure2d.png)

### Scatter plot of breast tissue smoothness vs. compactness 
![Figure 1e](plots/figure2e.png)

### Density plot of area means for both diagnosis
![Figure 1f](plots/figure2f.png)

## Part Three
### Cell-type ratio distributions
![Figure 2a](plots/Figure3a.png)

### Half-life vs alpha-life scatter
![Figure 2b](plots/Figure3b.png)

### Heatmap across cell types and time
Clustering genes across time in each sample. 
![Figure 2c](plots/Figure3c.png)

### Pathway enrichment heatmap
![Figure 2d](plots/Figure3d.png)

### Bubble plot of kinetic regimes
![Figure 2e](plots/Figure3e.png)

### Stacked proportions
![Figure 2f](plots/Figure3f.png)

### Directed cellâ€“cell interaction network
![Figure 2g](plots/Figure3g.png)
The edge width describes the intensity of relationships between teo nodes. 


